fix(ivy): add flag to skip non-exported classes

crisbeto opened the pull request:
In ViewEngine we were only generating code for exported classes, however with Ivy we do it no matter whether the class has been exported or not. These changes add an extra flag that allows consumers to opt into the ViewEngine behavior. The flag works by treating non-exported classes as if they're set to `jit: true`.

Fixes #33724.

crisbeto commented on packages/compiler-cli/src/transformers/api.ts:
```diff
@@ -392,6 +392,12 @@ export interface CompilerOptions extends ts.CompilerOptions {
+   * Whether the compiler should avoid generating code for classes that haven't been exported.
+   * This is only active when building with `enableIvy: true`.
+   */
+  skipNonExportedClasses?: boolean;
```
Some questions for the reviewer since the ticket wasn't very descriptive:

1. I'm open to suggestion on what we should call the flag.
2. Do we want to default it to `true` so it matches ViewEngine or leave it as an opt-in?

devversion replied:
That's a good question. I think there is still discussion going on. I'd vote for this being opt-in since long-term it feels like we need to come up with something that works better for unit testing w/ AOT. I've explained in the original Github issue.

alxhub replied:
In general, I think when a flag is `true` it should do something, not prevent something.

So instead of `skipNonExportedClasses`, let's go with `compileNonExportedClasses`, and make it default to `true`. Then if users don't want the behavior, they explicitly disable it, rather than explicitly enable skipping it.

alxhub commented on packages/compiler-cli/ngcc/src/analysis/decoration_analyzer.ts:
```diff
@@ -88,18 +88,19 @@ export class DecorationAnalyzer {
     new DirectiveDecoratorHandler(
         this.reflectionHost, this.evaluator, this.fullRegistry, NOOP_DEFAULT_IMPORT_RECORDER,
-        this.isCore),
+        this.isCore, /* skipNonExportedClasses */ false),
```
I think the default here should be `true`, as ngcc only processes code which was intended for View Engine.

alxhub commented on packages/compiler-cli/test/ngtsc/ngtsc_spec.ts:
```diff
@@ -3030,6 +3030,58 @@ runInEachFileSystem(os => {
       expect(jsContents).toMatch(/directives: \[DirA,\s+DirB\]/);
     });
 
+    it('should not emit directive definitions for non-exported classes if configured', () => {
```
Can we put these all in a `describe('non-exported classes', ...)`?

devversion commented:
This PR will close #33724. This issue also provides some context and a reproduction for later reference.

alxhub commented on packages/compiler-cli/src/ngtsc/annotations/src/ng_module.ts:
```diff
@@ -44,7 +44,8 @@ export class NgModuleDecoratorHandler implements DecoratorHandler<NgModuleAnalys
       private referencesRegistry: ReferencesRegistry, private isCore: boolean,
       private routeAnalyzer: NgModuleRouteAnalyzer|null, private refEmitter: ReferenceEmitter,
-      private defaultImportRecorder: DefaultImportRecorder, private localeId?: string) {}
+      private defaultImportRecorder: DefaultImportRecorder, private skipNonExportedClasses: boolean,
```
Rather than carry `skipNonExportedClasses` around to each of the `DecoratorHandlers` (and in fact this PR only does 2 of them - non-exported classes will still be processed as `@Injectable`s for example) we should nip this in the bud.

In `IvyCompilation` inside `analyze`, we can skip the class for all decorator handlers without passing the flag down into each one.

crisbeto commented:
The feedback has been addressed.

AndrewKushnir commented:
- [VE Presubmit](http://test/OCL:281782004:BASE:281782075:1574361287241:5df846bb)
- [Ivy Presubmit](https://test.corp.google.com/OCL:281782004:BASE:281782303:1574361333701:8982e1df)

alxhub commented on packages/compiler-cli/test/ngtsc/ngtsc_spec.ts:
```diff
@@ -5021,6 +5021,61 @@ export const Foo = Foo__PRE_R3__;
+
+    describe('non-exported classes', () => {
+      it('should not emit directive definitions for non-exported classes if configured', () => {
+        env.tsconfig({skipNonExportedClasses: true});
```
This can move into a `beforeEach` inside the `describe`. That centralizes the common thread between the tests.

crisbeto commented:
Updated based on the latest set of feedback.

alxhub commented on packages/compiler-cli/src/ngtsc/transform/src/compilation.ts:
```diff
@@ -80,7 +80,7 @@ export class IvyCompilation {
       private importRewriter: ImportRewriter, private incrementalState: IncrementalState,
       private perf: PerfRecorder, private sourceToFactorySymbols: Map<string, Set<string>>|null,
-      private scopeRegistry: LocalModuleScopeRegistry) {}
+      private scopeRegistry: LocalModuleScopeRegistry, private compileNonExportedClasses = true) {}
```
No default values in compiler code. Explicitly pass in the value at the callsite.

alxhub commented on packages/compiler-cli/src/ngtsc/program.ts:
```diff
@@ -607,7 +607,7 @@ export class NgtscProgram implements api.Program {
     return new IvyCompilation(
         handlers, this.reflector, this.importRewriter, this.incrementalState, this.perfRecorder,
-        this.sourceToFactorySymbols, scopeRegistry);
+        this.sourceToFactorySymbols, scopeRegistry, this.options.compileNonExportedClasses);
```
```suggestion
        this.sourceToFactorySymbols, scopeRegistry, this.options.compileNonExportedClasses !== false);
```

alxhub commented:
[Presubmit](http://test/OCL:281803190:BASE:281803243:1574366708702:92f3701e)
Ivy Presubmit: not needed, no a g3 behavior change.

crisbeto committed:
fix(ivy): add flag to skip non-exported classes

In ViewEngine we were only generating code for exported classes, however with Ivy we do it no matter whether the class has been exported or not. These changes add an extra flag that allows consumers to opt into the ViewEngine behavior. The flag works by treating non-exported classes as if they're set to `jit: true`.

Fixes #33724.

AndrewKushnir commented:
@crisbeto g3 presubmit looks good (I removed the "presubmit" label). Could you please rebase this PR and resolve a conflict? Thank you.

crisbeto commented:
Rebased.
